src/main/java/ch/heigvd/pro/b04/statistics/StatisticsController.java

```diff
@@ -206,42 +208,96 @@ public ServerPollStatistics statisticsForPoll(
-                .getIdParticipant()))
-            .entrySet()
-            .parallelStream()
-            // Get the latest vote for the participant.
-            .map(participantIdentifierListEntry ->
-                participantIdentifierListEntry.getValue().stream()
-                    .max(Comparator.comparing(serverVote -> serverVote.getIdVote().getTimeVote()))
-                    .map(ServerVote::isAnswerChecked)
-                    .orElse(false))
-            .collect(Collectors.toUnmodifiableList());
-
-        // Fetch the positive and negative answer counts.
-        answerBuilder.negative(((int) groupedByParticipant.stream().filter(b -> !b).count()));
-        answerBuilder.positive(((int) groupedByParticipant.stream().filter(b -> b).count()));
-
-        questionBuilder.answer(answerBuilder.build());
+        if (votes == null) {
+          votes = answer.getVoteSet();
+        } else {
+          votes.addAll(answer.getVoteSet());
+        }
       }
 
-      builder.question(questionBuilder.build());
-    }
+      var groupedByParticipant = votes
+          .parallelStream()
+          // Group the votes by participant.
+          .collect(Collectors.groupingBy(serverVote -> serverVote.getIdVote()
+              .getIdxParticipant()
+              .getIdParticipant()))
+          .entrySet()
+          .parallelStream()
+
+          // Sort votes by the most recent ones
+          .map(participantIdentifierListEntry ->
+              participantIdentifierListEntry.getValue().stream()
+                  .sorted((a1, a2) -> a1.compareTo(a2))
+                  //.limit(question.getAnswersMax())
+                  //.max(Comparator.comparing(serverVote -> serverVote.getIdVote().getTimeVote()))
+                  //.map(ServerVote::isAnswerChecked)
+                  .collect(Collectors.toList()))
+          //.orElse(false))
+          .collect(Collectors.toUnmodifiableList());
+      //at this point, groupedByParticipant is a List<List<ServerVote>>
+
+      //*2 to avoid surely out of bounds, theoretically (+1) would be enough
+      short[] positive = new short[question.getAnswersToQuestion().size() * 2];
+      short[] negative = new short[question.getAnswersToQuestion().size() * 2];
+
+      //counting negative and positives votes one participant at a time
+      for (short x = 0; x < groupedByParticipant.size(); ++x) {
+        short participantCount = 0;
+        short y = 0;
+        for (; y < groupedByParticipant.get(x).size(); ++y) {
+          if (groupedByParticipant.get(x).get(y).isAnswerChecked()) {
+            ++participantCount;
+            //limit to answersMax positive votes.
+            // remember most recent votes are the first ones in the list.
+            if (!(participantCount > question.getAnswersMax())) {
+              ++positive[(int) groupedByParticipant.get(x).get(y).getIdVote().getIdxServerAnswer()
+                  .getIdAnswer()
+                  .getIdAnswer()];
```

Comment:
Are we sure that this is correct ? The questions are not guaranteed to be assigned some identifiers starting at `0` and finishing at `n`.

Reply:
That's why the size*2

Reply:
But do we even have a guarantee that it fits in `size*2` ?

Reply:
Yes because identifiers begin at 1 and are incremented at each add. *2 is theoritically enough, we can do *3 to be sure